File: go/mechanisms/evm/erc4337/safe.go
package erc4337

import (
	"crypto/ecdsa"
	"fmt"
	"math/big"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
	"golang.org/x/crypto/sha3"
)

// SafeAccountConfig contains configuration for Safe smart account.
type SafeAccountConfig struct {
	// Owner is the owner's private key
	Owner *ecdsa.PrivateKey
	// ChainID is the chain ID
	ChainID int64
	// Salt for counterfactual address derivation
	Salt *big.Int
	// EntryPoint is the EntryPoint contract address
	EntryPoint common.Address
	// Threshold is the number of signatures required (default 1)
	Threshold int
}

// SafeSmartAccount implements a Safe-based smart account for ERC-4337.
type SafeSmartAccount struct {
	owner          *ecdsa.PrivateKey
	ownerAddress   common.Address
	chainID        int64
	salt           *big.Int
	entryPoint     common.Address
	threshold      int
	cachedAddress  common.Address
	cachedInitCode []byte
}

// NewSafeSmartAccount creates a new Safe smart account.
func NewSafeSmartAccount(config SafeAccountConfig) (*SafeSmartAccount, error) {
	if config.Owner == nil {
		return nil, fmt.Errorf("owner private key is required")
	}

	ownerAddress := crypto.PubkeyToAddress(config.Owner.PublicKey)

	salt := config.Salt
	if salt == nil {
		salt = big.NewInt(0)
	}

	entryPoint := config.EntryPoint
	if entryPoint == (common.Address{}) {
		entryPoint = common.HexToAddress(EntryPointV07Address)
	}

	threshold := config.Threshold
	if threshold <= 0 {
		threshold = 1
	}

	return &SafeSmartAccount{
		owner:        config.Owner,
		ownerAddress: ownerAddress,
		chainID:      config.ChainID,
		salt:         salt,
		entryPoint:   entryPoint,
		threshold:    threshold,
	}, nil
}

// GetAddress returns the counterfactual Safe address.
func (s *SafeSmartAccount) GetAddress() (common.Address, error) {
	if s.cachedAddress != (common.Address{}) {
		return s.cachedAddress, nil
	}

	// Calculate counterfactual address
	initCode, err := s.GetInitCode()
	if err != nil {
		return common.Address{}, err
	}

	// The address is derived from the proxy factory CREATE2
	// address = keccak256(0xff ++ factory ++ salt ++ keccak256(initCode))[12:]
	factoryAddress := Safe4337Addresses.ProxyFactory

	// Calculate salt with owner and setup
	saltHash := s.calculateSalt()

	// Calculate init code hash for proxy
	proxyInitCode := s.getProxyCreationCode()
	initCodeHash := crypto.Keccak256(proxyInitCode)

	// CREATE2 address calculation
	data := make([]byte, 1+20+32+32)
	data[0] = 0xff
	copy(data[1:21], factoryAddress.Bytes())
	copy(data[21:53], saltHash[:])
	copy(data[53:85], initCodeHash)

	hash := crypto.Keccak256(data)
	s.cachedAddress = common.BytesToAddress(hash[12:])
	s.cachedInitCode = initCode

	return s.cachedAddress, nil
}

// SignUserOpHash signs a UserOperation hash.
func (s *SafeSmartAccount) SignUserOpHash(userOpHash common.Hash) ([]byte, error) {
	// For Safe 4337, we sign the EIP-712 typed data
	// The Safe module expects: signature = abi.encodePacked(r, s, v)

	// Create message hash for signing
	messageHash := s.getSafeUserOpHash(userOpHash)

	// Sign with the owner key
	signature, err := crypto.Sign(messageHash[:], s.owner)
	if err != nil {
		return nil, fmt.Errorf("failed to sign: %w", err)
	}

	// Adjust v value for Ethereum (add 27)
	if signature[64] < 27 {
		signature[64] += 27
	}

	return signature, nil
}

// GetInitCode returns the init code for deploying the Safe.
func (s *SafeSmartAccount) GetInitCode() ([]byte, error) {
	if len(s.cachedInitCode) > 0 {
		return s.cachedInitCode, nil
	}

	// Init code = factory address + createProxyWithNonce(singleton, initializer, saltNonce)
	factoryAddress := Safe4337Addresses.ProxyFactory

	// Build the initializer data for Safe setup
	initializer := s.buildInitializer()

	// Encode createProxyWithNonce call
	// Function selector: 0x1688f0b9
	// createProxyWithNonce(address _singleton, bytes memory initializer, uint256 saltNonce)
	selector := []byte{0x16, 0x88, 0xf0, 0xb9}

	// ABI encode the parameters
	encoded := encodeCreateProxyWithNonce(Safe4337Addresses.Singleton, initializer, s.salt)

	initCode := make([]byte, 20+4+len(encoded))
	copy(initCode[0:20], factoryAddress.Bytes())
	copy(initCode[20:24], selector)
	copy(initCode[24:], encoded)

	s.cachedInitCode = initCode
	return initCode, nil
}

// IsDeployed checks if the account is deployed.
func (s *SafeSmartAccount) IsDeployed() (bool, error) {
	// This would require an RPC call to check code at address
	// For now, return false (caller should check via eth_getCode)
	return false, nil
}

// EncodeExecute encodes a call to Safe's execute function.
func (s *SafeSmartAccount) EncodeExecute(target common.Address, value *big.Int, data []byte) ([]byte, error) {
	// Safe 4337 module uses executeUserOp
	// Function: executeUserOp(address to, uint256 value, bytes data, uint8 operation)
	// Selector: 0x541d63c8
	selector := []byte{0x54, 0x1d, 0x63, 0xc8}

	// Operation 0 = CALL
	operation := uint8(0)

	encoded := encodeExecuteUserOp(target, value, data, operation)

	result := make([]byte, 4+len(encoded))
	copy(result[0:4], selector)
	copy(result[4:], encoded)

	return result, nil
}

// EncodeExecuteBatch encodes a batch call.
func (s *SafeSmartAccount) EncodeExecuteBatch(targets []common.Address, values []*big.Int, datas [][]byte) ([]byte, error) {
	if len(targets) != len(values) || len(targets) != len(datas) {
		return nil, fmt.Errorf("targets, values, and datas must have same length")
	}

	// For batch, we use Safe's multiSend
	// First encode individual transactions for multiSend
	var multiSendData []byte
	for i := range targets {
		// Each tx: operation (1 byte) + to (20 bytes) + value (32 bytes) + dataLength (32 bytes) + data
		txData := encodeMultiSendTx(targets[i], values[i], datas[i])
		multiSendData = append(multiSendData, txData...)
	}

	// Then encode call to multiSend contract via executeUserOp with DELEGATECALL
	// multiSend(bytes transactions)
	// Selector: 0x8d80ff0a
	multiSendSelector := []byte{0x8d, 0x80, 0xff, 0x0a}
	multiSendCall := encodeBytes(multiSendData)
	multiSendCalldata := make([]byte, 4+len(multiSendCall))
	copy(multiSendCalldata[0:4], multiSendSelector)
	copy(multiSendCalldata[4:], multiSendCall)

	// Execute via Safe 4337 module with DELEGATECALL (operation = 1)
	multiSendLib := common.HexToAddress("0x38869bf66a61cF6bDB996A6aE40D5853Fd43B526") // MultiSend v1.3.0

	selector := []byte{0x54, 0x1d, 0x63, 0xc8} // executeUserOp
	operation := uint8(1)                      // DELEGATECALL

	encoded := encodeExecuteUserOp(multiSendLib, big.NewInt(0), multiSendCalldata, operation)

	result := make([]byte, 4+len(encoded))
	copy(result[0:4], selector)
	copy(result[4:], encoded)

	return result, nil
}

// buildInitializer builds the Safe setup initializer data.
func (s *SafeSmartAccount) buildInitializer() []byte {
	// Safe.setup(
	//   address[] _owners,
	//   uint256 _threshold,
	//   address to,
	//   bytes data,
	//   address fallbackHandler,
	//   address paymentToken,
	//   uint256 payment,
	//   address paymentReceiver
	// )
	// Selector: 0xb63e800d

	selector := []byte{0xb6, 0x3e, 0x80, 0x0d}

	owners := []common.Address{s.ownerAddress}
	threshold := big.NewInt(int64(s.threshold))

	// to = AddModulesLib to enable 4337 module
	toAddress := Safe4337Addresses.AddModulesLib

	// data = enableModules([Safe4337Module])
	// enableModules(address[])
	// Selector: 0x610b5925 (wrong - actually 0xa3f4df7e for enableModules)
	// Actually we use addModuleWithSetup from AddModulesLib
	moduleSetupData := encodeEnableModules([]common.Address{Safe4337Addresses.Module})

	fallbackHandler := Safe4337Addresses.FallbackHandler
	paymentToken := common.Address{}
	payment := big.NewInt(0)
	paymentReceiver := common.Address{}

	encoded := encodeSetup(owners, threshold, toAddress, moduleSetupData, fallbackHandler, paymentToken, payment, paymentReceiver)

	result := make([]byte, 4+len(encoded))
	copy(result[0:4], selector)
	copy(result[4:], encoded)

	return result
}

// calculateSalt calculates the CREATE2 salt.
func (s *SafeSmartAccount) calculateSalt() [32]byte {
	// Salt = keccak256(keccak256(initializer) ++ saltNonce)
	initializer := s.buildInitializer()
	initHash := crypto.Keccak256(initializer)

	data := make([]byte, 32+32)
	copy(data[0:32], initHash)
	s.salt.FillBytes(data[32:64])

	var result [32]byte
	copy(result[:], crypto.Keccak256(data))
	return result
}

// getProxyCreationCode returns the proxy creation code.
func (s *SafeSmartAccount) getProxyCreationCode() []byte {
	// This is the SafeProxy creation code with singleton address
	// The actual bytecode would come from the Safe contracts
	// For now, return a simplified version
	singleton := Safe4337Addresses.Singleton.Bytes()

	// Proxy creation code pattern (simplified)
	// In production, this should match the actual Safe proxy bytecode
	code := make([]byte, 0)
	code = append(code, []byte{0x60, 0x20}...) // PUSH1 0x20
	// ... more bytecode
	code = append(code, singleton...)

	return code
}

// getSafeUserOpHash creates the Safe-specific user op hash for signing.
func (s *SafeSmartAccount) getSafeUserOpHash(userOpHash common.Hash) common.Hash {
	// Safe 4337 uses a specific format
	// For simple EOA owners, we just sign the userOpHash directly
	// The Safe module will verify using ecrecover

	// Hash with Ethereum signed message prefix
	prefix := []byte("\x19Ethereum Signed Message:\n32")
	data := make([]byte, len(prefix)+32)
	copy(data[0:len(prefix)], prefix)
	copy(data[len(prefix):], userOpHash[:])

	return common.BytesToHash(crypto.Keccak256(data))
}

// ABI encoding helpers

func encodeCreateProxyWithNonce(singleton common.Address, initializer []byte, saltNonce *big.Int) []byte {
	// (address, bytes, uint256)
	// offset for bytes is 96 (3 * 32)
	result := make([]byte, 0)

	// singleton address (padded to 32 bytes)
	result = append(result, common.LeftPadBytes(singleton.Bytes(), 32)...)

	// offset to initializer bytes
	result = append(result, common.LeftPadBytes(big.NewInt(96).Bytes(), 32)...)

	// saltNonce
	result = append(result, common.LeftPadBytes(saltNonce.Bytes(), 32)...)

	// initializer bytes
	result = append(result, encodeBytes(initializer)...)

	return result
}

func encodeExecuteUserOp(to common.Address, value *big.Int, data []byte, operation uint8) []byte {
	// (address, uint256, bytes, uint8)
	result := make([]byte, 0)

	// to address
	result = append(result, common.LeftPadBytes(to.Bytes(), 32)...)

	// value
	if value == nil {
		value = big.NewInt(0)
	}
	result = append(result, common.LeftPadBytes(value.Bytes(), 32)...)

	// offset to data bytes (4 * 32 = 128)
	result = append(result, common.LeftPadBytes(big.NewInt(128).Bytes(), 32)...)

	// operation
	result = append(result, common.LeftPadBytes([]byte{operation}, 32)...)

	// data bytes
	result = append(result, encodeBytes(data)...)

	return result
}

func encodeBytes(data []byte) []byte {
	// length + padded data
	length := big.NewInt(int64(len(data)))
	paddedLength := ((len(data) + 31) / 32) * 32

	result := make([]byte, 32+paddedLength)
	copy(result[0:32], common.LeftPadBytes(length.Bytes(), 32))
	copy(result[32:32+len(data)], data)

	return result
}

func encodeSetup(owners []common.Address, threshold *big.Int, to common.Address, data []byte, fallbackHandler, paymentToken common.Address, payment *big.Int, paymentReceiver common.Address) []byte {
	result := make([]byte, 0)

	// Calculate offsets
	// Fixed params: threshold (32) + to (32) + offset to data (32) + fallbackHandler (32) + paymentToken (32) + payment (32) + paymentReceiver (32)
	// = 224 bytes after owners array offset

	// offset to owners array (8 * 32 = 256)
	result = append(result, common.LeftPadBytes(big.NewInt(256).Bytes(), 32)...)

	// threshold
	result = append(result, common.LeftPadBytes(threshold.Bytes(), 32)...)

	// to
	result = append(result, common.LeftPadBytes(to.Bytes(), 32)...)

	// Calculate offset for data
	ownersEncodedLen := 32 + 32*len(owners) // length + addresses
	dataOffset := 256 + ownersEncodedLen
	result = append(result, common.LeftPadBytes(big.NewInt(int64(dataOffset)).Bytes(), 32)...)

	// fallbackHandler
	result = append(result, common.LeftPadBytes(fallbackHandler.Bytes(), 32)...)

	// paymentToken
	result = append(result, common.LeftPadBytes(paymentToken.Bytes(), 32)...)

	// payment
	if payment == nil {
		payment = big.NewInt(0)
	}
	result = append(result, common.LeftPadBytes(payment.Bytes(), 32)...)

	// paymentReceiver
	result = append(result, common.LeftPadBytes(paymentReceiver.Bytes(), 32)...)

	// owners array
	result = append(result, common.LeftPadBytes(big.NewInt(int64(len(owners))).Bytes(), 32)...)
	for _, owner := range owners {
		result = append(result, common.LeftPadBytes(owner.Bytes(), 32)...)
	}

	// data bytes
	result = append(result, encodeBytes(data)...)

	return result
}

func encodeEnableModules(modules []common.Address) []byte {
	// enableModules(address[])
	// Selector: 0xa3f4df7e
	selector := []byte{0xa3, 0xf4, 0xdf, 0x7e}

	// offset to array
	encoded := common.LeftPadBytes(big.NewInt(32).Bytes(), 32)

	// array length
	encoded = append(encoded, common.LeftPadBytes(big.NewInt(int64(len(modules))).Bytes(), 32)...)

	// array elements
	for _, module := range modules {
		encoded = append(encoded, common.LeftPadBytes(module.Bytes(), 32)...)
	}

	result := make([]byte, 4+len(encoded))
	copy(result[0:4], selector)
	copy(result[4:], encoded)

	return result
}

func encodeMultiSendTx(to common.Address, value *big.Int, data []byte) []byte {
	// operation (1) + to (20) + value (32) + dataLength (32) + data
	if value == nil {
		value = big.NewInt(0)
	}

	result := make([]byte, 1+20+32+32+len(data))
	result[0] = 0 // CALL operation

	copy(result[1:21], to.Bytes())

	valueBytes := common.LeftPadBytes(value.Bytes(), 32)
	copy(result[21:53], valueBytes)

	lengthBytes := common.LeftPadBytes(big.NewInt(int64(len(data))).Bytes(), 32)
	copy(result[53:85], lengthBytes)

	copy(result[85:], data)

	return result
}

// keccak256 helper
func keccak256(data []byte) []byte {
	h := sha3.NewLegacyKeccak256()
	h.Write(data)
	return h.Sum(nil)
}

File: typescript/packages/mechanisms/evm/src/erc4337/accounts/safe.ts
/**
 * Safe Smart Account for ERC-4337
 *
 * Implements SmartAccountSigner using Safe's 4337 module.
 * Supports:
 * - Single-owner and multi-sig configurations
 * - Counterfactual address computation
 * - Safe 4337 module v0.3.0
 *
 * @see https://docs.safe.global/advanced/erc-4337
 */

import type { Address, Hex, PublicClient, WalletClient } from "viem";
import {
  encodeFunctionData,
  encodeAbiParameters,
  concat,
  pad,
  toHex,
  keccak256,
  getContractAddress,
  hexToBytes,
} from "viem";
import type { SmartAccountSigner } from "../types.js";

/**
 * Safe 4337 module addresses (v0.3.0)
 * Deployed on all major EVM chains at the same addresses
 */
export const SAFE_4337_ADDRESSES = {
  /** Safe 4337 Module */
  module: "0xa581c4A4DB7175302464fF3C06380BC3270b4037" as Address,
  /** Safe Module Setup */
  moduleSetup: "0x2dd68b007B46fBe91B9A7c3EDa5A7a1063cB5b47" as Address,
  /** Safe Singleton */
  singleton: "0x29fcB43b46531BcA003ddC8FCB67FFE91900C762" as Address,
  /** Safe Proxy Factory */
  proxyFactory: "0x4e1DCf7AD4e460CfD30791CCC4F9c8a4f820ec67" as Address,
  /** Safe Fallback Handler */
  fallbackHandler: "0xfd0732Dc9E303f09fCEf3a7388Ad10A83459Ec99" as Address,
  /** Add Modules Lib */
  addModulesLib: "0x8EcD4ec46D4D2a6B64fE960B3D64e8B94B2234eb" as Address,
} as const;

/**
 * Safe Proxy Factory ABI (essential functions)
 */
const PROXY_FACTORY_ABI = [
  {
    inputs: [
      { name: "singleton", type: "address" },
      { name: "initializer", type: "bytes" },
      { name: "saltNonce", type: "uint256" },
    ],
    name: "createProxyWithNonce",
    outputs: [{ name: "proxy", type: "address" }],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      { name: "singleton", type: "address" },
      { name: "initializer", type: "bytes" },
      { name: "saltNonce", type: "uint256" },
    ],
    name: "proxyCreationCode",
    outputs: [{ name: "", type: "bytes" }],
    stateMutability: "view",
    type: "function",
  },
] as const;

/**
 * Safe Singleton ABI (essential functions)
 */
const SAFE_ABI = [
  {
    inputs: [
      { name: "owners", type: "address[]" },
      { name: "threshold", type: "uint256" },
      { name: "to", type: "address" },
      { name: "data", type: "bytes" },
      { name: "fallbackHandler", type: "address" },
      { name: "paymentToken", type: "address" },
      { name: "payment", type: "uint256" },
      { name: "paymentReceiver", type: "address" },
    ],
    name: "setup",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      { name: "to", type: "address" },
      { name: "value", type: "uint256" },
      { name: "data", type: "bytes" },
      { name: "operation", type: "uint8" },
    ],
    name: "execTransactionFromModule",
    outputs: [{ name: "success", type: "bool" }],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

/**
 * Add Modules Lib ABI
 */
const ADD_MODULES_LIB_ABI = [
  {
    inputs: [{ name: "modules", type: "address[]" }],
    name: "enableModules",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

/**
 * Safe 4337 Module ABI
 */
const SAFE_4337_MODULE_ABI = [
  {
    inputs: [
      { name: "to", type: "address" },
      { name: "value", type: "uint256" },
      { name: "data", type: "bytes" },
      { name: "operation", type: "uint8" },
    ],
    name: "executeUserOp",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      { name: "tos", type: "address[]" },
      { name: "values", type: "uint256[]" },
      { name: "datas", type: "bytes[]" },
      { name: "operations", type: "uint8[]" },
    ],
    name: "executeUserOpBatch",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

/**
 * Safe smart account configuration
 */
export interface SafeSmartAccountConfig {
  /** Wallet client for signing */
  signer: WalletClient;
  /** Public client for reading chain state */
  publicClient: PublicClient;
  /** Chain ID */
  chainId: number;
  /** Owner addresses (for multi-sig, defaults to signer address) */
  owners?: Address[];
  /** Threshold for multi-sig (defaults to 1) */
  threshold?: number;
  /** Salt nonce for address generation */
  saltNonce?: bigint;
  /** Custom Safe addresses (optional) */
  addresses?: Partial<typeof SAFE_4337_ADDRESSES>;
}

/**
 * Safe smart account implementing SmartAccountSigner
 */
export class SafeSmartAccount implements SmartAccountSigner {
  private readonly signer: WalletClient;
  private readonly publicClient: PublicClient;
  private readonly chainId: number;
  private readonly owners: Address[];
  private readonly threshold: number;
  private readonly saltNonce: bigint;
  private readonly addresses: typeof SAFE_4337_ADDRESSES;

  private cachedAddress?: Address;
  private cachedInitCode?: Hex;
  private deploymentChecked = false;
  private isAccountDeployed = false;

  constructor(config: SafeSmartAccountConfig) {
    this.signer = config.signer;
    this.publicClient = config.publicClient;
    this.chainId = config.chainId;
    this.threshold = config.threshold ?? 1;
    this.saltNonce = config.saltNonce ?? 0n;
    this.addresses = {
      ...SAFE_4337_ADDRESSES,
      ...config.addresses,
    };

    // Default to signer address if no owners provided
    if (config.owners && config.owners.length > 0) {
      this.owners = config.owners;
    } else if (config.signer.account?.address) {
      this.owners = [config.signer.account.address];
    } else {
      throw new Error("Either owners or signer with account must be provided");
    }

    if (this.threshold > this.owners.length) {
      throw new Error("Threshold cannot be greater than number of owners");
    }
  }

  /**
   * Get the smart account address (counterfactual)
   */
  async getAddress(): Promise<Address> {
    if (this.cachedAddress) {
      return this.cachedAddress;
    }

    const initCode = await this.getInitCode();

    // Extract initializer from init code
    const initializerData = `0x${initCode.slice(2 + 40 * 2)}` as Hex;

    // Compute counterfactual address
    const salt = keccak256(
      encodeAbiParameters(
        [{ type: "bytes32" }, { type: "uint256" }],
        [keccak256(initializerData), this.saltNonce],
      ),
    );

    // Get proxy creation code
    const proxyCreationCode = await this.publicClient.readContract({
      address: this.addresses.proxyFactory,
      abi: PROXY_FACTORY_ABI,
      functionName: "proxyCreationCode",
      args: [this.addresses.singleton, initializerData, this.saltNonce],
    }) as Hex;

    // Compute CREATE2 address
    this.cachedAddress = getContractAddress({
      bytecode: proxyCreationCode,
      from: this.addresses.proxyFactory,
      opcode: "CREATE2",
      salt,
    });

    return this.cachedAddress;
  }

  /**
   * Sign a UserOperation hash
   */
  async signUserOpHash(userOpHash: Hex): Promise<Hex> {
    if (!this.signer.account) {
      throw new Error("Signer account not available");
    }

    // Sign the hash with EIP-712 formatted signature for Safe
    const signature = await this.signer.signMessage({
      account: this.signer.account,
      message: { raw: hexToBytes(userOpHash) },
    });

    // Format signature for Safe (add signature type byte)
    // Type 0: EOA signature (most common)
    return concat([signature, "0x00"]) as Hex;
  }

  /**
   * Get the account's init code for deployment
   */
  async getInitCode(): Promise<Hex> {
    // Check if already deployed
    if (await this.isDeployed()) {
      return "0x" as Hex;
    }

    if (this.cachedInitCode) {
      return this.cachedInitCode;
    }

    // Build Safe setup data with 4337 module
    const setupModulesData = encodeFunctionData({
      abi: ADD_MODULES_LIB_ABI,
      functionName: "enableModules",
      args: [[this.addresses.module]],
    });

    const safeSetupData = encodeFunctionData({
      abi: SAFE_ABI,
      functionName: "setup",
      args: [
        this.owners,
        BigInt(this.threshold),
        this.addresses.addModulesLib, // to: AddModulesLib
        setupModulesData, // data: enableModules([module])
        this.addresses.fallbackHandler,
        "0x0000000000000000000000000000000000000000" as Address, // paymentToken
        0n, // payment
        "0x0000000000000000000000000000000000000000" as Address, // paymentReceiver
      ],
    });

    // Build factory call data
    const createProxyData = encodeFunctionData({
      abi: PROXY_FACTORY_ABI,
      functionName: "createProxyWithNonce",
      args: [this.addresses.singleton, safeSetupData, this.saltNonce],
    });

    // Init code = factory address + factory call data
    this.cachedInitCode = concat([
      this.addresses.proxyFactory,
      createProxyData,
    ]) as Hex;

    return this.cachedInitCode;
  }

  /**
   * Check if the account is deployed
   */
  async isDeployed(): Promise<boolean> {
    if (this.deploymentChecked) {
      return this.isAccountDeployed;
    }

    const address = this.cachedAddress ?? await this.getAddress();
    const code = await this.publicClient.getCode({ address });

    this.deploymentChecked = true;
    this.isAccountDeployed = code !== undefined && code !== "0x";

    return this.isAccountDeployed;
  }

  /**
   * Encode a call to the account's execute function
   */
  encodeExecute(target: Address, value: bigint, data: Hex): Hex {
    return encodeFunctionData({
      abi: SAFE_4337_MODULE_ABI,
      functionName: "executeUserOp",
      args: [
        target,
        value,
        data,
        0, // operation: CALL
      ],
    });
  }

  /**
   * Encode a batch call to the account's executeBatch function
   */
  encodeExecuteBatch(
    targets: Address[],
    values: bigint[],
    datas: Hex[],
  ): Hex {
    if (targets.length !== values.length || targets.length !== datas.length) {
      throw new Error("Array lengths must match");
    }

    const operations = targets.map(() => 0); // All CALL operations

    return encodeFunctionData({
      abi: SAFE_4337_MODULE_ABI,
      functionName: "executeUserOpBatch",
      args: [targets, values, datas, operations],
    });
  }

  /**
   * Get the counterfactual address without caching
   */
  async getCounterfactualAddress(): Promise<Address> {
    return this.getAddress();
  }

  /**
   * Get the account's nonce from EntryPoint
   */
  async getNonce(entryPoint: Address, key = 0n): Promise<bigint> {
    const address = await this.getAddress();

    const nonce = await this.publicClient.readContract({
      address: entryPoint,
      abi: [
        {
          inputs: [
            { name: "sender", type: "address" },
            { name: "key", type: "uint192" },
          ],
          name: "getNonce",
          outputs: [{ name: "nonce", type: "uint256" }],
          stateMutability: "view",
          type: "function",
        },
      ],
      functionName: "getNonce",
      args: [address, key],
    });

    return nonce as bigint;
  }

  /**
   * Get the Safe's owners
   */
  getOwners(): Address[] {
    return [...this.owners];
  }

  /**
   * Get the Safe's threshold
   */
  getThreshold(): number {
    return this.threshold;
  }

  /**
   * Clear cached values (useful after deployment)
   */
  clearCache(): void {
    this.cachedAddress = undefined;
    this.cachedInitCode = undefined;
    this.deploymentChecked = false;
    this.isAccountDeployed = false;
  }
}

/**
 * Create a Safe smart account
 */
export function createSafeSmartAccount(
  config: SafeSmartAccountConfig,
): SafeSmartAccount {
  return new SafeSmartAccount(config);
}
